metronome/checkpointing/service/src/io/iohk/metronome/checkpointing/service/models/Block.scala

```diff
@@ -0,0 +1,62 @@
+package io.iohk.metronome.checkpointing.service.models
+
+import io.iohk.metronome.checkpointing.interpreter.models.Transaction
+
+/** Represents what the HotStuff paper called "nodes" as the "tree",
+  * with the transactions in the body being the "commands".
+  *
+  * The block contents are specific to the checkpointing application.
+  *
+  * The header and body are separated because headers have to part
+  * of the Checkpoint Certificate; there's no need to repeat all
+  * the transactions there, the Merkle root will make it possible
+  * to prove that a given CheckpointCandidate transaction was
+  * indeed part of the block. The headers are needed for parent-child
+  * validation in the certificate as well.
+  */
+sealed abstract case class Block(
+    header: Block.Header,
+    body: Block.Body
+) {
+  def hash: Block.Header.Hash = header.hash
+}
+
+object Block {
+
+  /** Create a from a header and body we received from the network.
+    * It will need to be validated before it can be used, to make sure
+    * the header really belongs to the body.
+    */
+  def makeUnsafe(header: Header, body: Body): Block =
+    new Block(header, body) {}
+
+  /** Create a block from a header and a body, updating the `bodyHash` in the
+    * header to make sure the final block hash is valid.
+    */
+  def make(
+      header: Header,
+      body: Body
+  ) = makeUnsafe(header = header.copy(bodyHash = body.hash), body = body)
+
+  case class Header(
+      parentHash: Header.Hash,
+      // Hash of the Ledger before executing the block.
+      preStateHash: Ledger.Hash,
```

Comment:
Why do we need preStateHash, isn't preStateHash = parent.postStateHash?

Reply:
It is, in fact I added a [better constructor](https://github.com/input-output-hk/metronome/pull/20/files#diff-b97700721336d64d734806933d3aa6a8e188fd821718570e27f71df253791a35R35-R50) to `Block` that takes the parent and just copies this. 

We don't need it, probably, I added because I once worked on a project that had it like this. I'll remove it, we can always add it if it's useful.